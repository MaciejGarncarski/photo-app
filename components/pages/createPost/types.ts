import { z } from 'zod';

import { PostDetailsSchema } from './CreatePost';

export type PostDetails = z.infer<typeof PostDetailsSchema>;

type FinalImage = {
  id: string;
  file: Blob | null;
};
export type FinalImages = Array<FinalImage | undefined>;
