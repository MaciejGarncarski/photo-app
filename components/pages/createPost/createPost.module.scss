@use '/styles/common.scss';

.createPost {
  display: grid;
  width: clamp(20rem, 70%, 60rem);
  margin: 2rem auto;
  margin-bottom: 4rem;
  gap: 2rem;
}

.actionButtons {
  display: flex;
  justify-content: space-between;
}

.textAreaContainer {
  border: 1px solid var(--color-primary);
  border-radius: var(--radius-small);
  position: relative;

  @include common.mediaquery('desktop') {
    display: flex;
  }
}

.label {
  position: absolute;
  background-color: var(--bg-secondary);
}

.loadingContainer {
  display: flex;
  flex-direction: column;
}

.addPhoto {
  display: grid;
  width: 100%;
}