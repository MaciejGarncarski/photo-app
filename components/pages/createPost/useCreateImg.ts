import { RefObject, useEffect } from 'react';
import { PixelCrop } from 'react-image-crop';

import { canvasPreview } from '@/utils/canvasPreview';

type UseCreateImg = {
  completedCrop?: PixelCrop;
  imgRef: RefObject<HTMLImageElement>;
  previewCanvasRef: RefObject<HTMLCanvasElement>;
  setFinalImg: (image: Blob | null) => void;
  isError: boolean;
};

export const useCreateImg = ({ completedCrop, imgRef, previewCanvasRef, setFinalImg, isError }: UseCreateImg) => {
  useEffect(() => {
    const createImg = async () => {
      if (!imgRef.current || !previewCanvasRef.current || !completedCrop?.height || !completedCrop.width) {
        return;
      }

      if (isError) {
        setFinalImg(null);
      }

      const { toBlob } = await canvasPreview(imgRef.current, previewCanvasRef.current, completedCrop);
      setFinalImg(await toBlob);
    };
    createImg();
  }, [completedCrop, imgRef, isError, previewCanvasRef, setFinalImg]);
};
