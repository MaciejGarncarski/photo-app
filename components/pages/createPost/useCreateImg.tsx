import { RefObject, useEffect } from 'react';
import { PixelCrop } from 'react-image-crop';

import { canvasPreview } from '@/utils/canvasPreview';

type UseCreateImg = {
  completedCrop?: PixelCrop;
  imgRef: RefObject<HTMLImageElement>;
  previewCanvasRef: RefObject<HTMLCanvasElement>;
  setFinalImg: (image: Blob) => void;
};

export const useCreateImg = ({
  completedCrop,
  imgRef,
  previewCanvasRef,
  setFinalImg,
}: UseCreateImg) => {
  useEffect(() => {
    const createImg = async () => {
      if (
        !imgRef.current ||
        !previewCanvasRef.current ||
        !completedCrop?.height ||
        !completedCrop.width
      ) {
        return;
      }
      const { toBlob } = await canvasPreview(
        imgRef.current,
        previewCanvasRef.current,
        completedCrop
      );

      setFinalImg(await toBlob);
    };
    createImg();
  }, [completedCrop, imgRef, previewCanvasRef, setFinalImg]);
};
