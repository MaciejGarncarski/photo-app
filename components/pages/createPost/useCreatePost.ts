import { useMutation } from '@tanstack/react-query';
import axios from 'axios';
import { useRouter } from 'next/router';

import { useAuth } from '@/components/organisms/signIn/useAuth';

type MutationValues = {
  description: string;
  image: Blob;
};

type UploadPostArguments = {
  author: string;
} & MutationValues;

const uplaodPost = async ({ description, image, author }: UploadPostArguments) => {
  return await axios.putForm<unknown, null, UploadPostArguments>('/api/post', {
    description,
    image,
    author,
  });
};

export const useCreatePost = () => {
  const { push } = useRouter();
  const { session } = useAuth();

  return useMutation(
    async ({ description, image }: MutationValues) => {
      if (!session?.user?.id) {
        return;
      }
      return await uplaodPost({ author: session.user.id, description, image });
    },
    {
      onSuccess: () => {
        push('/');
      },
    },
  );
};
