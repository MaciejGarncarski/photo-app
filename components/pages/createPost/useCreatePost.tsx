import { useMutation } from '@tanstack/react-query';
import axios from 'axios';
import { useSession } from 'next-auth/react';

type MutationValues = {
  description: string;
  location: string;
  image: Blob;
};

type UploadPostArguments = {
  author: string;
} & MutationValues;

const uplaodPost = async ({ description, location, image, author }: UploadPostArguments) => {
  return await axios.putForm<null, null, UploadPostArguments>('/api/post', {
    description,
    location,
    image,
    author,
  });
};

export const useCreatePost = () => {
  const { data: session } = useSession();
  return useMutation(async ({ description, location, image }: MutationValues) => {
    if (!session?.user?.id) {
      return;
    }
    return await uplaodPost({ author: session.user.id, description, image, location });
  });
};
