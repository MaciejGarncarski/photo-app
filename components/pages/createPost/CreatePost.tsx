/* eslint-disable @next/next/no-img-element */
import { zodResolver } from '@hookform/resolvers/zod';
import { useRouter } from 'next/router';
import { NextSeo } from 'next-seo';
import { useState } from 'react';
import { SubmitHandler, useForm } from 'react-hook-form';
import { toast } from 'react-hot-toast';
import { v4 } from 'uuid';
import { z } from 'zod';

import { useAuth } from '@/hooks/useAuth';

import { Button } from '@/components/atoms/button/Button';
import { CreatePostItemContainer } from '@/components/atoms/createPostItemContainer/CreatePostItemContainer';
import { Heading } from '@/components/atoms/heading/Heading';
import { LoadingHeading } from '@/components/atoms/loadingHeading/LoadingHeading';
import { ModalContainer } from '@/components/atoms/modal/ModalContainer';
import { useModal } from '@/components/atoms/modal/useModal';
import { TextArea } from '@/components/atoms/textArea/TextArea';
import { AspectRatioButtons } from '@/components/molecules/aspectRatioButtons/AspectRatioButtons';
import { ConfirmationAlert } from '@/components/molecules/confirmationAlert/ConfirmationAlert';
import { CropImage } from '@/components/molecules/cropImage/CropImage';
import { ImagesPreview } from '@/components/molecules/imagesPreview/ImagesPreview';
import { useConvertToBase64 } from '@/components/pages/createPost/useConvertToBase64';
import { useSendNewPost } from '@/components/pages/createPost/useSendNewPost';
import { useUploadImage } from '@/components/pages/createPost/useUploadImage';

import styles from './createPost.module.scss';

const PostDetailsSchema = z.object({
  description: z.string().max(200, { message: 'Maximum characters exceeded' }),
});

type PostDetails = z.infer<typeof PostDetailsSchema>;

type FinalImage = {
  id: number;
  file: Blob | null;
};
export type FinalImages = Array<FinalImage | undefined>;

export type ImagesBase64 = Array<
  | {
      id: number;
      src: string;
    }
  | undefined
>;

export const CreatePost = () => {
  const router = useRouter();
  const { session } = useAuth();

  const [isCropping, setIsCropping] = useState<boolean>(false);

  const [aspectRatio, setAspectRatio] = useState<number>(1);
  const [finalImages, setFinalImages] = useState<FinalImages>([]);
  const [finalImagesBase64, setFinalImagesBase64] = useState<ImagesBase64>();

  useConvertToBase64(finalImages, setFinalImagesBase64);

  const { open, close, modalOpen } = useModal();

  const uploadImage = useUploadImage();
  const sendNewPost = useSendNewPost();

  const {
    register,
    handleSubmit,
    formState: { dirtyFields, errors },
  } = useForm<PostDetails>({
    resolver: zodResolver(PostDetailsSchema),
    defaultValues: {
      description: '',
    },
  });

  const onSubmit: SubmitHandler<PostDetails> = async ({ description }) => {
    if (!finalImages[0]?.file) {
      return;
    }

    const uuid = v4();
    const folder = `${session?.user?.id}/posts/${uuid}`;

    const images = await Promise.all(
      finalImages.map(async (image) => {
        if (!image?.file) {
          return;
        }
        return await uploadImage.mutateAsync(
          { imageBlob: image.file, folder },
          {
            onError: () => {
              toast.error('Error');
            },
          },
        );
      }),
    );

    const imageUrls = images.filter((img): img is string => !!img);

    await sendNewPost.mutateAsync(
      { description, imageUrls },
      {
        onSuccess: async () => {
          await router.push('/');
        },
        onError: () => {
          toast.error('Could not add post.');
        },
      },
    );
  };

  const handleRemoveImage = (id: number) => {
    const filteredState = finalImages.filter((finalImg) => {
      return finalImg?.id !== id;
    });
    setFinalImages(filteredState);
  };

  if (uploadImage.isLoading || sendNewPost.isLoading || sendNewPost.isSuccess) {
    return <LoadingHeading headingText="Uploading your post." />;
  }

  const isSubmitDisabled = !dirtyFields.description || finalImages.length === 0;

  return (
    <section aria-labelledby="Create new post" className={styles.createPost}>
      <NextSeo title="Create new post" />

      {finalImages.length <= 3 && (
        <div className={styles.addPhoto}>
          <CreatePostItemContainer>
            <CropImage
              isCropping={isCropping}
              setIsCropping={setIsCropping}
              setFinalImages={setFinalImages}
              finalImages={finalImages}
              aspectRatio={aspectRatio}
            />
          </CreatePostItemContainer>
        </div>
      )}
      {isCropping && <AspectRatioButtons aspect={aspectRatio} setAspect={setAspectRatio} />}
      {finalImagesBase64 && <ImagesPreview imagesBase64={finalImagesBase64} onRemove={handleRemoveImage} />}
      <form onSubmit={handleSubmit(onSubmit)}>
        <CreatePostItemContainer>
          <Heading tag="h2">Info about post</Heading>
          <TextArea label="description" {...register('description')} error={errors.description} />
          <div className={styles.actionButtons}>
            <Button variant="secondary" onClick={open}>
              Cancel
            </Button>
            <Button type="submit" disabled={isSubmitDisabled}>
              Complete
            </Button>
          </div>
        </CreatePostItemContainer>
      </form>

      <ModalContainer>
        {modalOpen && <ConfirmationAlert headingText="Cancel?" close={close} onConfirm={() => router.push('/')} />}
      </ModalContainer>
    </section>
  );
};
