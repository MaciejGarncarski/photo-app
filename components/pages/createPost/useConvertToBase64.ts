import { useEffect } from 'react';

import { FinalImages, ImagesBase64 } from '@/components/pages/createPost/CreatePost';

export const useConvertToBase64 = (images: FinalImages, setImagesBase64: (finalImg: ImagesBase64) => void) => {
  useEffect(() => {
    const convertedFiles = images.map((finalImg) => {
      if (finalImg?.file) {
        const img = URL.createObjectURL(finalImg?.file);
        return {
          id: finalImg?.id,
          src: img,
        };
      }
    });
    setImagesBase64(convertedFiles);
  }, [images, setImagesBase64]);
};
