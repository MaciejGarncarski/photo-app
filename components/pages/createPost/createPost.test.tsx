import { CreatePost } from '@/components/pages/createPost/CreatePost';

import { render, screen } from '@/tests';

jest.mock('next/router', () => require('next-router-mock'));

jest.mock('next-auth/react', () => {
  const originalModule = jest.requireActual('next-auth/react');
  const mockSession = {
    expires: new Date(Date.now() + 2 * 86400).toISOString(),
    user: { username: 'admin' },
  };
  return {
    __esModule: true,
    ...originalModule,
    useSession: jest.fn(() => {
      return { data: mockSession, status: 'authenticated' };
    }),
  };
});

describe('<CreatePost />', () => {
  beforeEach(() => {
    render(<CreatePost />);
  });

  it('CreatePost component should render', async () => {
    const container = await screen.findByTestId(/createPost-container/i);
    expect(container).toBeInTheDocument();
  });

  it('button should be disabled when no description provided', async () => {
    const button = await screen.findByText(/complete/i);
    expect(button).toHaveAttribute('disabled', '');
    expect(button).toMatchSnapshot();
  });
});
