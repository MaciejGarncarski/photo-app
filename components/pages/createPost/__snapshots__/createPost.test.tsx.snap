// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatePost /> button should be disabled when no description provided 1`] = `
<button
  class="button"
  disabled=""
  type="submit"
>
  Complete
</button>
`;
